### PR TITLE
[JSDK-2648] Add example for enabling and disabling tracks

### PR DESCRIPTION
**Pull request by PikaJoyce**
WIP : Still need to work on helper functions, need to fix up the bug, need to clean up code, move things around but generally most of it is functional. 

Description : Adding the example for enabling and disabling tracks

UI : 
Screen Shot 2020-01-31 at 2 33 36 PM
buttons

**Contributing to Twilio**

> All third-party contributors acknowledge that any contributions they provide will be made under the same open-source license that the open-source project is provided under.

- [x] I acknowledge that all my contributions will be made under the project's license.

### CODE REVIEW

File: examples/enablingdisablingtracks/src/helpers.js [hunk 1]
```diff
@@ -0,0 +1,47 @@
+'use strict';
```
Comment (manjeshbhargav) [inline examples/enablingdisablingtracks/src/helpers.js]: @PikaJoyce You should also add the code to unmute tracks and listen for enabled/disabled events here:
```js
/**
 * Mute/unmute your media in a Room.
 * @param {Room} room - The Room you have joined
 * @param {'audio'|'video'} kind - The type of media you want to mute/unmute
 * @param {'mute'|'unmute'} action - Whether you want to mute/unmute
 */
function muteOrUnmuteYourMedia(room, kind, action) {
  const publications = kind === 'audio'
    ? room.localParticipant.audioTracks
    : room.localParticipant.videoTracks;

  publications.forEach(function(publication) {
    if (action === 'mute') {
      publication.track.disable();
    } else {
      publication.track.enable();
    }
  });
}

/**
 * Mute your audio in a Room.
 * @param {Room} room - The Room you have joined
 * @returns {void}
 */
function muteYourAudio(room) {
  muteOrUnmuteYourMedia(room, 'audio', 'mute');
}

/**
 * Mute your video in a Room.
 * @param {Room} room - The Room you have joined
 * @returns {void}
 */
function muteYourVideo(room) {
  muteOrUnmuteYourMedia(room, 'video', 'mute');
}

/**
 * Unmute your audio in a Room.
 * @param {Room} room - The Room you have joined
 * @returns {void}
 */
function unmuteYourAudio(room) {
  muteOrUnmuteYourMedia(room, 'audio', 'unmute');
}

/**
 * Unmute your video in a Room.
 * @param {Room} room - The Room you have joined
 * @returns {void}
 */
function unmuteYourVideo(room) {
  muteOrUnmuteYourMedia(room, 'video', 'unmute');
}

/**
 * A RemoteParticipant muted or unmuted its media.
 * @param {Room} room - The Room you have joined
 * @param {function} onMutedMedia - Called when a RemoteParticipant muted its media
 * @param {function} onUnmutedMedia - Called when a RemoteParticipant unmuted its media
 * @returns {void}
 */
function participantMutedOrUnmutedMedia(room, onMutedMedia, onUnmutedMedia) {
  room.on('trackSubscribed', function(track, publication, participant) {
    track.on('disabled', function() {
      onMutedMedia(track, participant);
    });
    track.on('enabled', function() {
      onUnmutedMedia(track, participant);
    });
  });
}

exports.muteYourAudio = muteYourAudio;
exports.muteYourVideo = muteYourVideo;
exports.muteYourAudio = unmuteYourAudio;
exports.muteYourVideo = unmuteYourVideo;
exports.participantMutedOrUnmutedMedia = participantMutedOrUnmutedMedia;
```

File: examples/enablingdisablingtracks/src/index.js [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+(async function(){
+  // Load the code snippet.
+  const snippet = await getSnippet('./helpers.js');
+  const pre = document.querySelector('pre.language-javascript');
+  
+  pre.innerHTML = Prism.highlight(snippet, Prism.languages.javascript);
+
+  // Get the credentials to connect to the Room.
+  const credsP1 = await getRoomCredentials();
+
```
Comment (manjeshbhargav) [inline examples/enablingdisablingtracks/src/index.js]:
```suggestion
```

File: examples/enablingdisablingtracks/src/index.js [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+  // Create room instance and name for participants to join.
+  const roomP1 = await Video.connect(credsP1.token);
+
+  // Set room name for participant 2 to join.
+  roomName = roomP1.name;
+
+  // Connecting remote participants.
+  const roomP2 = await Video.connect(credsP2.token, {
+    name:roomName,
```
Comment (manjeshbhargav) [inline examples/enablingdisablingtracks/src/index.js]:
```suggestion
    name: roomName,
```

File: examples/enablingdisablingtracks/src/index.js [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+    tracks: [],
```
Comment (manjeshbhargav) [inline examples/enablingdisablingtracks/src/index.js]:
```suggestion
    tracks: []
```

File: examples/enablingdisablingtracks/src/index.js [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+  });
+
+  // Muting audio track and video tracks
+  muteAudioBtn.onclick = () => {
+    const mute = muteAudioBtn.innerHTML === '<img src="./icons/volume-mute-fill.svg"> Mute Audio';
+    const localUser = roomP1.localParticipant;
+
+    muteAudio(localUser, mute);
+
+    if(mute) {
```
Comment (manjeshbhargav) [inline examples/enablingdisablingtracks/src/index.js]: @PikaJoyce This should happen in the "enabled" / "disabled" callbacks.

File: examples/enablingdisablingtracks/src/index.js [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+    });
+  }));
+
+  // Disconnect from the Room 
+  window.onbeforeunload = () => {
+    roomP1.disconnect();
+    roomP2.disconnect();
+    roomName = null;
+  }
+}());
```
Comment (manjeshbhargav) [inline examples/enablingdisablingtracks/src/index.js]: @PikaJoyce Here, you can use do something like:
```js
participantMutedOrUnmutedMedia(roomP2, track => {
  if (track.kind === 'audio') {
    audioPreview.appendChild(track.attach())
  } else if (track.kind === 'video') {
    videoPreview.appendChild(track.attach())
  }
}, track => {
  track.detach().forEach(element => {
    element.remove();
  });
});
```

File: examples/enablingdisablingtracks/src/index.js [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+    }
+  }
+
+  muteVideoBtn.onclick = () => {
+    const mute = muteVideoBtn.innerHTML === '<img src="./icons/pause-fill.svg"> Mute Video';
+    const localUser = roomP1.localParticipant;
+
+    muteVideo(localUser, mute);
+
+    if(mute) {
```
Comment (manjeshbhargav) [inline examples/enablingdisablingtracks/src/index.js]: @PikaJoyce This should happen in the "enabled" / "disabled" callbacks.

File: server/index.js [hunk 8]
```diff
@@ -32,7 +32,9 @@ var app = express();
   'mediadevices',
   'networkquality',
   'reconnection',
-  'screenshare'
+  'screenshare',
+  'enablingdisablingtracks'
```
Comment (manjeshbhargav) [inline server/index.js]:
```suggestion
  'localmediacontrols'
```

File: package.json [hunk 9]
```diff
@@ -16,6 +16,7 @@
     "build:examples:reconnection": "copyfiles -f examples/reconnection/src/helpers.js examples/reconnection/public && browserify examples/reconnection/src/index.js > examples/reconnection/public/index.js",
     "build:examples:screenshare": "copyfiles -f examples/screenshare/src/helpers.js examples/screenshare/public && browserify examples/screenshare/src/index.js > examples/screenshare/public/index.js",
     "build:quickstart": "browserify quickstart/src/index.js > quickstart/public/index.js",
+    "build:examples:enablingdisablingtracks": "copyfiles -f examples/enablingdisablingtracks/src/helpers.js examples/enablingdisablingtracks/public && browserify examples/enablingdisablingtracks/src/index.js > examples/enablingdisablingtracks/public/index.js",
```
Comment (manjeshbhargav) [inline package.json]:
```suggestion
    "build:examples:localmediacontrols": "copyfiles -f examples/localmediacontrols/src/helpers.js examples/localmediacontrols/public && browserify examples/localmediacontrols/src/index.js > examples/localmediacontrols/public/index.js",
```

File: package.json [hunk 10]
```diff
@@ -27,6 +28,7 @@
     "clean:examples:networkquality": "rimraf examples/networkquality/public/index.js examples/networkquality/public/helpers.js",
     "clean:examples:reconnection": "rimraf examples/reconnection/public/index.js examples/reconnection/public/helpers.js",
     "clean:examples:screenshare": "rimraf examples/screenshare/public/index.js examples/screenshare/public/helpers.js",
+    "clean:examples:enablingdisablingtracks": "rimraf examples/enablingdisablingtracks/public/index.js examples/enablingdisablingtracks/public/helpers.js",
```
Comment (manjeshbhargav) [inline package.json]:
```suggestion
    "clean:examples:localmediacontrols": "rimraf examples/localmediacontrols/public/index.js examples/localmediacontrols/public/helpers.js",
```

File: examples/enablingdisablingtracks/src/index.js (hunk 6 above)
Comment (PikaJoyce) [reply examples/enablingdisablingtracks/src/index.js]: Having some issues with this method. Track is successfully disabled and enabled but the video track won't detach from the DOM.

Comment (PikaJoyce) [reply examples/enablingdisablingtracks/src/index.js]: Solved!

**Comment by PikaJoyce** [comment]
Fixed CSS a little bit so it's a little better looking!
Screen Shot 2020-02-04 at 1 04 44 PM

File: .gitignore [hunk 11]
```diff
@@ -68,3 +69,4 @@ examples/mediadevices/public/helpers.js
 examples/networkquality/public/helpers.js
 examples/reconnection/public/helpers.js
 examples/screenshare/public/helpers.js
+examples/localmediacontrols/public/helpers.js
```
Comment (manjeshbhargav) [inline .gitignore]: @PikaJoyce Please add a new line here:
```suggestion
examples/localmediacontrols/public/helpers.js

```

File: examples/index.html [hunk 12]
```diff
@@ -157,6 +157,22 @@ <h4 class="card-header">
...
+        <div class="card-deck">
+          <div class="card" style="text-decoration: none">
+            <a href="/localmediacontrols">
+              <h4 class="card-header">
+                Enabling and Disabling Tracks
+              </h4>
+            </a>
+            <div class="card-block">
+              <p class="card-text">
+                This app demonstrates how to enable/disable tracks and handle the events on the remote side.
```
Comment (manjeshbhargav) [inline examples/index.html]:
```suggestion
                This app demonstrates how to enable/disable your LocalTracks and handle the events on the remote side.
```

File: examples/localmediacontrols/src/index.js [hunk 13]
```diff
@@ -0,0 +1,137 @@
...
+
+  // Connecting remote participants.
+  const roomP2 = await Video.connect(credsP2.token, {
+    name: roomName,
+    tracks: []
+  });
+
+  // Muting audio track and video tracks click handlers
+  muteAudioBtn.onclick = () => {
+    const mute = muteAudioBtn.style.display === 'flex';
```
Comment (manjeshbhargav) [inline examples/localmediacontrols/src/index.js]: @PikaJoyce 
```suggestion
```

File: examples/localmediacontrols/src/index.js [hunk 14]
```diff
@@ -0,0 +1,137 @@
...
+
+    muteYourAudio(roomP1);
+
+    if (mute) {
```
Comment (manjeshbhargav) [inline examples/localmediacontrols/src/index.js]:
```suggestion
```

File: examples/localmediacontrols/src/index.js [hunk 15]
```diff
@@ -0,0 +1,137 @@
...
+      muteAudioBtn.style.display = 'none';
+      unmuteAudioBtn.style.display = 'flex';
+    } else {
```
Comment (manjeshbhargav) [inline examples/localmediacontrols/src/index.js]:
```suggestion
```

File: examples/localmediacontrols/src/index.js [hunk 16]
```diff
@@ -0,0 +1,137 @@
...
+      muteAudioBtn.style.display = 'flex';
```
Comment (manjeshbhargav) [inline examples/localmediacontrols/src/index.js]:
```suggestion
```

File: examples/localmediacontrols/src/index.js [hunk 17]
```diff
@@ -0,0 +1,137 @@
...
+      unmuteAudioBtn.style.display = 'none';
```
Comment (manjeshbhargav) [inline examples/localmediacontrols/src/index.js]:
```suggestion
```

File: examples/localmediacontrols/src/index.js [hunk 18]
```diff
@@ -0,0 +1,137 @@
...
+    }
```
Comment (manjeshbhargav) [inline examples/localmediacontrols/src/index.js]:
```suggestion
```

File: examples/localmediacontrols/src/index.js [hunk 19]
```diff
@@ -0,0 +1,137 @@
...
+  }
+
+  unmuteAudioBtn.onclick = () => {
+    const mute = !unmuteAudioBtn.style.display === 'flex';
```
Comment (manjeshbhargav) [inline examples/localmediacontrols/src/index.js]:
```suggestion
```

File: examples/localmediacontrols/src/index.js [hunk 20]
```diff
@@ -0,0 +1,137 @@
...
+
+    unmuteYourAudio(roomP1);
+
+    if (!mute) {
+      unmuteAudioBtn.style.display = 'none';
+      muteAudioBtn.style.display = 'flex';
+    } else {
```
Comment (manjeshbhargav) [inline examples/localmediacontrols/src/index.js]:
```suggestion
```

File: examples/localmediacontrols/src/index.js [hunk 21]
```diff
@@ -0,0 +1,137 @@
...
+      unmuteAudioBtn.style.display = 'flex';
```
Comment (manjeshbhargav) [inline examples/localmediacontrols/src/index.js]:
```suggestion
```

File: examples/localmediacontrols/src/index.js [hunk 22]
```diff
@@ -0,0 +1,137 @@
...
+      muteAudioBtn.style.display = 'none';
```
Comment (manjeshbhargav) [inline examples/localmediacontrols/src/index.js]:
```suggestion
```

File: examples/localmediacontrols/src/index.js [hunk 23]
```diff
@@ -0,0 +1,137 @@
...
+    }
```
Comment (manjeshbhargav) [inline examples/localmediacontrols/src/index.js]:
```suggestion
```

File: examples/localmediacontrols/src/index.js [hunk 24]
```diff
@@ -0,0 +1,137 @@
...
+
+    muteYourVideo(roomP1);
+
+    if (mute) {
+      muteVideoBtn.style.display = 'none';
+      unmuteVideoBtn.style.display = 'flex';
+    } else {
+      muteVideoBtn.style.display = 'flex';
+      unmuteVideoBtn.style.display = 'none';
+    }
```
Comment (manjeshbhargav) [inline examples/localmediacontrols/src/index.js]: @PikaJoyce You don't need to handle both cases here:
```js
muteYourVideo(roomP1);
muteVideoBtn.style.display = 'none';
unmuteVideoBtn.style.display = 'flex';
```

File: examples/localmediacontrols/src/index.js [hunk 25]
```diff
@@ -0,0 +1,137 @@
...
+    unmuteYourVideo(roomP1);
+
+    if (!mute) {
+      unmuteVideoBtn.style.display = 'none';
+      muteVideoBtn.style.display = 'flex';
+    } else {
+      unmuteVideoBtn.style.display = 'flex';
+      muteVideoBtn.style.display = 'none';
+    }
+  }
```
Comment (manjeshbhargav) [inline examples/localmediacontrols/src/index.js]: @PikaJoyce You don't need to handle both cases here:
```js
unmuteYourVideo(roomP1);
unmuteVideoBtn.style.display = 'none';
muteVideoBtn.style.display = 'flex';
```

File: examples/localmediacontrols/src/index.js [hunk 26]
```diff
@@ -0,0 +1,137 @@
...
+    })};
+  
+  const attach = track => {
+    if (track.kind === 'audio') {
+      audioPreview.appendChild(track.attach())
+    }
+    if (track.kind === 'video') {
+      videoPreview.appendChild(track.attach())
+    }
+  };
```
Comment (manjeshbhargav) [inline examples/localmediacontrols/src/index.js]: @PikaJoyce Since you are calling these callbacks only once, you can define them inline in `participantMutedOrUnmutedMedia()`.

File: examples/localmediacontrols/src/index.js [hunk 27]
```diff
@@ -0,0 +1,137 @@
...
+    participantMutedOrUnmutedMedia(roomP2, detach, attach);
+}));
+
+  // Disconnect from the Room 
+  window.onbeforeunload = () => {
+    roomP1.disconnect();
+    roomP2.disconnect();
+    roomName = null;
+  }
+}());
```
Comment (manjeshbhargav) [inline examples/localmediacontrols/src/index.js]: @PikaJoyce Add extra line here:
```suggestion
}());

```

File: examples/localmediacontrols/src/index.js (hunk 26 above)
Comment (PikaJoyce) [reply examples/localmediacontrols/src/index.js]: Sounds good!

**Comment by PikaJoyce** [comment]
localmedia

File: examples/localmediacontrols/src/index.js [hunk 28]
```diff
@@ -125,13 +84,23 @@ let roomName = null;
       }
     }
 
-    participantMutedOrUnmutedMedia(roomP2, detach, attach);
-}));
+    participantMutedOrUnmutedMedia(roomP2, track => {
+      track.detach().forEach(element => {
+        element.remove();
+      })}, track => {
```
Comment (manjeshbhargav) [inline examples/localmediacontrols/src/index.js]:
```suggestion
      });
    }, track => {
```
